tests/kernel/wave/wave_e2e_test.py

```diff
@@ -180,3 +181,104 @@ def test(
     ):
         test(a, b)
         assert_allclose(a.T, b)
+
+
+@require_e2e
+def test_im2col():
+    # TODO: we don't support unaligned access at the moment so all sizes must
+    # be aligned to WG/Wave sizes, c * hw * wf == 8 and number of windows == 64
+    n, c, h, w = 1, 2, 9, 9  # Image.
+    cf, hf, wf = c, 2, 2  # Filters.
+    padding = 0
+    stride = 1
+
+    sym = tkl.sym
+    ADDRESS_SPACE = sym.ADDRESS_SPACE
+
+    N, C, H, W = sym.N, sym.C, sym.H, sym.W
+    NF, HF, WF = sym.NF, sym.HF, sym.WF
+
+    H_OUT = (H + 2 * padding - HF) // stride + 1
+    W_OUT = (W + 2 * padding - WF) // stride + 1
+    SZ_OUT = H_OUT * W_OUT
+
+    # K = HF * WF * C
+    # M = SZ_OUT * N
+    M = sym.M
+    K = sym.K
+
+    # We unroll K dimension according to ELEMS_PER_THREAD value.
+    # i.e. for K==8 we will have 2 vector.gather's.
+    # Each WG will process 64 windows.
+    wave_size = 64
+    BLOCK_K = hf * wf * c
+    BLOCK_M = 64
+    ELEMS_PER_THREAD = 4
+
+    i = tkw.IndexMapping.iterator(0)
+    j = tkw.IndexMapping.iterator(1)
+
+    mapping = tkw.IndexMapping(
+        num_iterators=2,
+        inputs={
+            N: i // SZ_OUT,
+            C: j // (HF * WF),
+            H: (i % SZ_OUT) // W_OUT * stride + (j % (HF * WF)) // WF,
+            W: (i % SZ_OUT) % W_OUT * stride + (j % (HF * WF)) % WF,
+        },
+        outputs={M: i, K: j},
+    )
+
+    constraints: list[tkw.Constraint] = [
+        tkw.HardwareConstraint(
+            threads_per_wave=wave_size,
+            waves_per_block=(1, 1, 1),
+            vector_shapes={M: BLOCK_M, K: ELEMS_PER_THREAD},
+        )
+    ]
+    constraints += [tkw.WorkgroupConstraint(M, BLOCK_M, 0)]
+    constraints += [tkw.WorkgroupConstraint(K, BLOCK_K, 1)]
+    constraints += [tkw.WaveConstraint(M, BLOCK_M)]
+    constraints += [tkw.WaveConstraint(K, BLOCK_K)]

```

Comment:
does WaveConstraint do unexpected behavior if undefined as opposed to erroring out?

Reply:
Without any K constraints example is still compiles successfully and K access is still unrolled according to `vector_shapes`, but it gets indices weird: K index is always 0.